### PR TITLE
Fix tests for vision models

### PR DESCRIPTION
**Pull request by qubvel**
# What does this PR do?

Fixing tests for vision models

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

**Comment by qubvel** [comment]
run-slow: beit, detr, dinov2, vit, textnet

**Comment by MHRDYN7** [comment]
Hi @qubvel, I'm not sure what this draft pr intends to do, however it might be relevant to the pr #35138. That pr fixes the incompatibility of FlaxDinov2 with batch sizes of more than 1. This error could not be detected by the flax tests (same as the pytorch tests), when I first contributed this model. All the slow tests in transformers simply pass a single image with a batch size of 1 and that is why such batch sizes incompatibility errors might not be detected. As a result, I changed the images batch size to 2 for the flax dinov2 slow tests (in that pr, not yet merged). Probably, doing the same for all the other future model slow tests would greatly assist the development process. Also, may I request a review on pr #35138 so that FlaxDinov2 can be used properly.

**Comment by qubvel** [comment]
run-slow: beit, detr, dinov2, vit, textnet

**Comment by qubvel** [comment]
run-slow: beit, detr, dinov2, vit, textnet

**Comment by qubvel** [comment]
run-slow: beit, detr, dinov2, vit, textnet

**Comment by qubvel** [comment]
run-slow: beit, data2vec, dpt

**Comment by qubvel** [comment]
run-slow: detr

**Comment by qubvel** [comment]
run-slow: beit, detr, dinov2, vit, textnet, data2vec, dpt

**Comment by qubvel** [comment]
run-slow: beit, detr, dinov2, vit, textnet, data2vec, dpt

**Comment by qubvel** [comment]
run-slow: beit, data2vec, dpt, zoedepth

File: tests/models/beit/test_modeling_beit.py [hunk 1]
```diff
@@ -764,19 +764,13 @@ def test_inference_interpolate_pos_encoding(self):
         inputs = processor(images=image, return_tensors="pt", size={"height": 480, "width": 480})
         pixel_values = inputs.pixel_values.to(torch_device)
 
-        # with interpolate_pos_encoding being False an exception should be raised with higher resolution
-        # images than what the model supports.
-        self.assertFalse(processor.do_center_crop)
-        with torch.no_grad():
-            with self.assertRaises(ValueError, msg="doesn't match model"):
-                model(pixel_values, interpolate_pos_encoding=False)
-
```
Comment (qubvel) [inline tests/models/beit/test_modeling_beit.py:773]: We always interpolate, error raising was removed for ZoeDepth in
https://github.com/huggingface/transformers/pull/30136/files#diff-3f84bebd6be8d9c0f5c5068199f5c49eac8489d5fa466fb6fa08b0365e78dba4

File: src/transformers/models/beit/modeling_beit.py [hunk 2]
```diff
@@ -211,14 +216,11 @@ def forward(
             embeddings = embeddings * (1 - w) + mask_tokens * w
 
         cls_tokens = self.cls_token.expand(batch_size, -1, -1)
-        if self.position_embeddings is not None:
-            if interpolate_pos_encoding:
-                cls_tokens = cls_tokens + self.interpolate_pos_encoding(embeddings, height, width)
```
Comment (qubvel) [inline src/transformers/models/beit/modeling_beit.py:216]: This is actually a bug, but we probably never reach it because `self.position_embeddings` is None

File: src/transformers/models/beit/modeling_beit.py [hunk 3]
```diff
@@ -248,11 +250,7 @@ def __init__(self, config):
 
         self.projection = nn.Conv2d(num_channels, hidden_size, kernel_size=patch_size, stride=patch_size)
 
-    def forward(
-        self,
-        pixel_values: torch.Tensor,
-        position_embedding: Optional[torch.Tensor] = None,
-    ) -> torch.Tensor:
```
Comment (qubvel) [inline src/transformers/models/beit/modeling_beit.py:255]: for consistency with other models `position_embedding` removed from `BeitPatchEmbeddings` and applied in `BeitEmbeddings` module, which is a breaking change, but I suppose `BeitPatchEmbeddings` module is used only as a part of the `BeitEmbeddings`.

File: src/transformers/models/beit/modeling_beit.py (hunk 2 above)
Comment (ydshieh) [reply src/transformers/models/beit/modeling_beit.py:216]: Or it's likely no one never use `interpolate_pos_encoding=True`?

If it is `True`, would it fail the call at this point or it's just compute a different value of `cls_tokens`?

BTW, I see

```
        if config.use_absolute_position_embeddings:
            self.position_embeddings = nn.Parameter(torch.zeros(1, num_patches + 1, config.hidden_size))
        else:
            self.position_embeddings = None
```
So it's possible `self.position_embeddings` is not `None` if `config.use_absolute_position_embeddings` is `True?`

File: src/transformers/models/beit/modeling_beit.py [hunk 4]
```diff
@@ -196,12 +197,16 @@ def forward(
...
     ) -> torch.Tensor:
+        if self.position_embeddings is not None and interpolate_pos_encoding is not None:
+            warnings.warn(
+                "`interpolate_pos_encoding` argument has no effect for BEiTEmbeddings, embeddings are always "
+                "interpolated to the input image size. The argument will be removed in transformers v4.51.0."
+            )
+
         _, _, height, width = pixel_values.shape
-        embeddings, (patch_height, patch_width) = self.patch_embeddings(
-            pixel_values, self.position_embeddings[:, 1:, :] if self.position_embeddings is not None else None
```
Comment (ydshieh) [inline src/transformers/models/beit/modeling_beit.py:203]: It's possible that `self.position_embeddings` is not `None` right? See my comment below

**Review by ydshieh — COMMENTED** [review]
Before looking at the tests, I have some question in the modeling code changes 🙏

File: src/transformers/models/beit/modeling_beit.py (hunk 2 above)
Comment (qubvel) [reply src/transformers/models/beit/modeling_beit.py:216]: Yeah, I checked some popular models on the hub, all have `use_absolute_position_embeddings: false` (didn't do extensive testing tbh). So it's likely the combination of factors: no one uses `use_absolute_position_embeddings=True` (should be a new model) + `interpolate_pos_encoding=True`.

It's a bug introduced with adding `interpolate_pos_encoding` flag, it should be a `embedding = embedding + ..` not `cls_tokens`. But even in that case we will have double interpolatio: here and one in BeitPatchEmbedding, so I just cleaned this up.

Comment (ydshieh) [reply src/transformers/models/beit/modeling_beit.py:216]: OK, indeed. Thanks!

File: tests/models/beit/test_modeling_beit.py [hunk 5]
```diff
@@ -764,19 +764,13 @@ def test_inference_interpolate_pos_encoding(self):
...
         # with interpolate_pos_encoding being True the model should process the higher resolution image
         # successfully and produce the expected output.
         with torch.no_grad():
             outputs = model(pixel_values, interpolate_pos_encoding=True)
 
-        expected_shape = torch.Size((1, 1801, 768))
+        # 1 + (480 / 16) * (480 / 16) = 1 + 30 * 30 = 901
```
Comment (ydshieh) [inline tests/models/beit/test_modeling_beit.py:772]: Would it possible to have also the variable names that contain the numbers here? Like

```python
# var_name 1 + (var_name 2 / var_name 3) ...
# 1 + (480 / 16) * (480 / 16) = 1 + 30 * 30 = 901
```

or whatever format that could better to understand what those values from

File: tests/models/textnet/test_modeling_textnet.py [hunk 6]
```diff
@@ -327,14 +327,18 @@ def test_inference_no_head(self):
...
+                [0.0000, 1.7415, 1.2660],
+                [0.0000, 1.0084, 1.9692],
+                [0.0000, 1.7464, 1.7892],
+            ],
             device=torch_device,
         )
         torch.testing.assert_close(
-            output.feature_maps[-1][0][10][12][:10], expected_slice_backbone, rtol=1e-3, atol=1e-3
+            output.last_hidden_state[0, 12, :3, :3], expected_slice_backbone, rtol=1e-2, atol=1e-2
         )
```
Comment (ydshieh) [inline tests/models/textnet/test_modeling_textnet.py:342]: So I guess `logits` and `feature_maps` doesn't make sense here, so we change to other things to check the values?

File: tests/models/zoedepth/test_modeling_zoedepth.py [hunk 7]
```diff
@@ -300,8 +300,8 @@ def check_target_size(
             out_l_reduced = torch.nn.functional.interpolate(
                 out_l.unsqueeze(0).unsqueeze(1), size=img.size[::-1], mode="bicubic", align_corners=False
             )
-            self.assertTrue((np.array(out_l.shape)[::-1] == np.array(img.size) * 2).all())
```
Comment (ydshieh) [inline tests/models/zoedepth/test_modeling_zoedepth.py:303]: looks we remove one check here? As long as you believe this check is not necessary, good for me.

File: src/transformers/models/beit/modeling_beit.py [hunk 8]
```diff
@@ -196,12 +197,16 @@ def forward(
         self,
         pixel_values: torch.Tensor,
         bool_masked_pos: Optional[torch.BoolTensor] = None,
-        interpolate_pos_encoding: bool = False,
+        interpolate_pos_encoding: Optional[bool] = None,
     ) -> torch.Tensor:
+        if self.position_embeddings is not None and interpolate_pos_encoding is not None:
+            warnings.warn(
```
Comment (ydshieh) [inline src/transformers/models/beit/modeling_beit.py:203]: I kind remember, for logic inside `forward`, we would rather using `logger.warning_once()` ... to avoid a lot of spam log message (because `forward` is usually called many times).

WDYT?

**Review by ydshieh — APPROVED** [review]
LGTM, thanks a lot.

A nit question regarding `logger.warning_once()` or `warnings.warn`.

File: tests/models/textnet/test_modeling_textnet.py (hunk 6 above)
Comment (qubvel) [reply tests/models/textnet/test_modeling_textnet.py:342]: There is no logit actually, it's either `last_hidden_state` for `TextModel` or `feature_maps` for `TextBackbone`

File: tests/models/zoedepth/test_modeling_zoedepth.py (hunk 7 above)
Comment (qubvel) [reply tests/models/zoedepth/test_modeling_zoedepth.py:303]: this check compares shapes, I believe `torch.testing.assert_close` takes this into account as well

File: src/transformers/models/beit/modeling_beit.py (hunk 8 above)
Comment (qubvel) [reply src/transformers/models/beit/modeling_beit.py:203]: hmm, no problem switching to logger, but `warning.warn` also warn just once for the same message, a simple reproducer:

```python
import warnings

for i in range(10):
    warnings.warn("test")
```

output
```
/home/ubuntu/projects/transformers/test_warnings.py:4: UserWarning: test
  warnings.warn("test")
```

Comment (ydshieh) [reply src/transformers/models/beit/modeling_beit.py:203]: oh, ok!

File: tests/models/beit/test_modeling_beit.py (hunk 5 above)
Comment (qubvel) [inline tests/models/beit/test_modeling_beit.py:772]:
```suggestion
        # num_cls_tokens + (height / patch_size) * (width / patch_size)
        # 1 + (480 / 16) * (480 / 16) = 1 + 30 * 30 = 901
```

File: tests/models/data2vec/test_modeling_data2vec_vision.py [hunk 9]
```diff
@@ -565,17 +565,11 @@ def test_inference_interpolate_pos_encoding(self):
...
-            with self.assertRaises(ValueError, msg="doesn't match model"):
-                model(pixel_values, interpolate_pos_encoding=False)
-
         # with interpolate_pos_encoding being True the model should process the higher resolution image
         # successfully and produce the expected output.
         with torch.no_grad():
             outputs = model(pixel_values, interpolate_pos_encoding=True)
 
-        expected_shape = torch.Size((1, 1801, 768))
+        # 1 + (480 / 16) * (480 / 16) = 901
```
Comment (qubvel) [inline tests/models/data2vec/test_modeling_data2vec_vision.py:573]:
```suggestion
        # num_cls_tokens + (height / patch_size) * (width / patch_size)
        # 1 + (480 / 16) * (480 / 16) = 901
```

**Comment by qubvel** [comment]
run-slow: beit, data2vec, dpt, zoedepth, detr, dinov2, vit, textnet

**Review by ArthurZucker — APPROVED** [review]
🤗 thanks for taking care of our Ci's health!
